### PR TITLE
Add --allow-branch option to restrict publish to designated branches

### PR DESCRIPTION
**Pull request by FaHeymann**
## Description
As suggested in https://github.com/lerna/lerna/issues/1007 this introduces an `allow-branch` flag to the `publish` command.

## Motivation and Context
As above, I would like to use that feature too.

## How Has This Been Tested?
Test suite & manual in another project

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project. (I hope)
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. (I hope)
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by evocateur** [comment]
whoops, sorry, the integration tests seem to be borked :/

File: package.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
     "eslint-plugin-flowtype": "^2.35.0",
     "eslint-plugin-node": "^5.1.1",
     "file-url": "^2.0.2",
-    "jest": "^19.0.2",
+    "jest": "^21.1.0",
```
Comment (evocateur) [inline package.json]: oh, yeah, please don't change this? It's being held back for a reason (I haven't had time to figure out why jest 20+ breaks)

**Comment by FaHeymann** [comment]
I guess https://github.com/facebook/jest/issues/3584 is what is causing it, right? So I will rollback the jest version

**Comment by evocateur** [comment]
basically, yeah. I haven't had time to knit the integration tests back into the regular tests (as well as make them take less time)

File: README.md [hunk 2]
```diff
@@ -433,6 +433,28 @@ to certain guidelines, such as projects which use [commitizen](https://github.co
...
+{
+    "command": {
+        "publish": {
+          "allowBranch": "master"
+        }
+    }
+}
+```
+
+and you are not on the branch `my-new-feature` currently lerna will prevent you from publishing. To force a publish anyways you can use
```
Comment (evocateur) [inline README.md]: This should read
```
and you are not on the branch `master`, lerna will prevent you from publishing.
To force a publish despite this config, pass the `--allow-branch` flag:
```
and so on

File: src/commands/PublishCommand.js [hunk 3]
```diff
@@ -206,6 +214,12 @@ export default class PublishCommand extends Command {
         throw new Error("Detached git HEAD, please checkout a branch to publish changes.");
       }
 
+      const currentBranch = GitUtilities.getCurrentBranch(this.execOpts);
+      if (this.options.allowBranch && !minimatch(currentBranch, this.options.allowBranch)) {
+        throw new Error('Branch ' + currentBranch + ' is not allowed to be published. ' +
```
Comment (evocateur) [inline src/commands/PublishCommand.js]: Should be able to use `dedent` with multi-line template strings, here.

File: README.md (hunk 2 above)
Comment (FaHeymann) [reply README.md]: 👍 fixed

**Comment by FaHeymann** [comment]
@evocateur could you take a look again? I hope I fixed everything :)

**Comment by evocateur** [comment]
Looks great!

**Pull request merged**